### DISCUSSION TITLE
Removing KHR_mesh_quantization

### DISCUSSION THREAD

**Discussion by donmccurdy** [donmccurdy/glTF-Transform#233] [Q&A] [answered]
Asked in the three.js repo, how can you remove the `KHR_mesh_quantization` extension from a model?

**Comment by donmccurdy** [comment, marked as answer]
Should be something like...

```typescript
// remove_quantization.js
const { NodeIO } = require('@gltf-transform/core');
const { KHRONOS_EXTENSIONS } = require('@gltf-transform/extensions');

const io = new NodeIO().registerExtensions(KHRONOS_EXTENSIONS);
const document = io.read('input.glb');
const root = document.getRoot();
const tmpAttr = document.createAccessor('TMP');

// Replace int8/int16/uint8/uint16 attributes with float32.
for (const mesh of root.listMeshes()) {
	for (const prim of mesh.listPrimitives()) {
		for (const semantic of prim.listSemantics()) {

			const attr = prim.getAttribute(semantic);
			if (attr.getComponentSize() === 4) continue;
			if (semantic.startsWith('JOINTS_')) continue;

			// Create identical tmp attribute with float32.
			tmpAttr.copy(attr)
				.setNormalized(false)
				.setArray(new Float32Array(attr.getArray().length));

			// Copy data to tmp attribute.
			for (let i = 0; i < attr.getCount(); i++) {
				tmpAttr.setElement(i, attr.getElement(i, []));
			}

			// Merge changes back to original attribute.
			attr.copy(tmpAttr);

		}
	}
}

tmpAttr.dispose();

// Remove KHR_mesh_quantization.
root.listExtensionsUsed()
	.find((ext) => ext.extensionName === 'KHR_mesh_quantization')
	.dispose();

io.write('output.glb', document);
```

And then run the script with `node remove_quantization.js`.

If quantization was applied with certain tools, the model might have `KHR_texture_transform` listed under extensions with `gltf-transform inspect model.glb`. If this UV transform works as expected already there is no need to remove it. If you need to remove the UV transform as well, it would need to be applied to the UVs, not shown above.

~~If some of the attributes are normalized (this will print a warning in the code above) there may be some extra conversions needed; I'm not sure if that is commonly the case.~~

**Comment by carstenschwede** [comment]
Thanks for the code. I actually do have `KHR_mesh_quantization, KHR_texture_transform, KHR_materials_pbrSpecularGlossiness` in my particular model and I also get the warning `May need to apply normalization ratio!`.

The resulting model has some objects at very different scales in comparison to the original. Any idea?

**Reply by donmccurdy** [reply]
Ok – I've made some edits above to handle the normalization. `KHR_texture_transform` should be fine as long as your textures are appearing correctly on the model.

**Reply by carstenschwede** [reply]
Thanks a lot for the help. The output.glb model created using your script looks good (left). But exporting it using GLTFExporter and reimporting it again fails (right).

GLTFExporter shows the following warnings:

- `GLTFExporter: Creating normalized normal attribute from the non-normalized one.`
- `GLTFExporter: Attribute "skinIndex" converted to type UNSIGNED_SHORT.`

<img width="494" alt="Screenshot 2021-04-20 at 13 02 37" src="https://user-images.githubusercontent.com/234960/115385685-b369df00-a1d8-11eb-985a-2e15233caaa3.png">

**Reply by donmccurdy** [reply]
Updated with another guess — if that's not working and you'd be able to share the model (my email address is on https://github.com/donmccurdy) I can try debugging that too.  The first warning looks OK.

**Reply by carstenschwede** [reply]
Awesome, thanks a lot. Looks good now!
